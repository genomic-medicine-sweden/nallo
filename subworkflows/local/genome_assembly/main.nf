include { CAT_FASTQ } from '../../../modules/nf-core/cat/fastq/main'
include { HIFIASM   } from '../../../modules/nf-core/hifiasm'
include { YAK_COUNT } from '../../../modules/nf-core/yak/count/main'
include { GFASTATS  } from '../../../modules/nf-core/gfastats/main'

// This subworkflow assembles and outputs haplotypes from a set of reads, using hifiasm and gfastats.
// It assumes that no sample exists twice in different families.
workflow GENOME_ASSEMBLY {

    take:
    ch_reads      // channel: [ val(meta), fastq ]
    trio_binning  //    bool: Should we use trio binning mode where possible?

    main:
    ch_versions = Channel.empty()

    if (!trio_binning) {
        ch_reads
            .map { meta, fastq ->
                [ groupKey(meta, meta.n_files), fastq ]
            }
            .groupTuple()
            .multiMap { meta, reads ->
                reads : [ meta, reads, [] ]
                yak   : [ [], [], [] ]
            }
            .set { ch_hifiasm_in }

    } else {
        // First, we need to branch the samples based on their relationship
        ch_reads
            .branch { meta, _reads ->
                def is_parent = meta.relationship in ['father', 'mother']
                paired_parents             : is_parent && meta.has_other_parent
                unpaired_parents           : is_parent && !meta.has_other_parent
                children_with_both_parents : meta.relationship == 'child' && meta.two_parents
                children_with_single_parent: meta.relationship == 'child' && !meta.two_parents
                unknown                    : meta.relationship == 'unknown'
            }
            .set { ch_branched_samples }

        // Then, the files from parents of children with both parents will need to be concatenated before yak
        // in case there are multiple files for the same parent.
        ch_branched_samples.paired_parents
            .map { meta, fastq ->
                [ groupKey(meta, meta.n_files), fastq ]
            }
            .groupTuple()
            .branch { _meta, fastqs ->
                cat: fastqs.size() > 1
                no_cat: fastqs.size() == 1
            }
            .set { ch_paired_parents_for_yak }

        CAT_FASTQ (
            ch_paired_parents_for_yak.cat
        )
        ch_versions = ch_versions.mix(CAT_FASTQ.out.versions)

        YAK_COUNT (
            CAT_FASTQ.out.reads.concat(ch_paired_parents_for_yak.no_cat)
        )
        ch_versions = ch_versions.mix(YAK_COUNT.out.versions)

        YAK_COUNT.out.yak
            // Because a parent can have multiple children, and meta.children is a list of all children,
            // we need to return one tuple per child.
            .flatMap { meta, yak ->
                (meta.children ?: []).collect { child_id ->
                    [child_id, meta, yak]
                }
            }
            .branch { child_id, meta, yak ->
                paternal: meta.relationship == 'father'
                    return [ child_id, yak ]
                maternal: meta.relationship == 'mother'
                    return [ child_id, yak ]
            }
            .set { ch_yak_output }

        // Creates the input for trio-binned assemblies (children with both parents)
        ch_branched_samples.children_with_both_parents
            .map { meta, fastq ->
                [ groupKey(meta, meta.n_files), fastq ]
            }
            .groupTuple()
            .map { meta, reads -> [ meta.id, meta, reads ] }
            .join(ch_yak_output.paternal)
            .join(ch_yak_output.maternal)
            .map { _id, meta, reads, yak_paternal, yak_maternal ->
                [ meta, reads, yak_paternal, yak_maternal ]
            }
            .set { ch_with_both_parents }

        // Creates the input for the non-trio binned assemblies
        // (children with a single parent, parents, and unknown samples)
        ch_branched_samples.children_with_single_parent
            .concat(ch_branched_samples.paired_parents)
            .concat(ch_branched_samples.unpaired_parents)
            .concat(ch_branched_samples.unknown)
            .map { meta, fastq ->
                [ groupKey(meta, meta.n_files), fastq ]
            }
            .groupTuple()
            .map { meta, fastqs ->
                [ meta, fastqs, [], [] ]
            }
            .set { ch_other_cases }

        // Combines the two, and creates the input for hifiasm
        ch_with_both_parents
            .concat(ch_other_cases)
            .multiMap { meta, reads, yak_paternal, yak_maternal ->
                reads : [ meta, reads, []                  ]
                yak   : [ meta, yak_paternal, yak_maternal ]
            }
            .set { ch_hifiasm_in }
    }

    HIFIASM (
        ch_hifiasm_in.reads,
        ch_hifiasm_in.yak,
        [[],[],[]],
        [[],[]]
    )
    ch_versions = ch_versions.mix(HIFIASM.out.versions)

    HIFIASM.out.hap1_contigs
        .map { meta, fasta -> [ meta + [ 'haplotype': 1 ], fasta ] }
        .set { ch_gfastats_paternal_in }

    HIFIASM.out.hap2_contigs
        .map { meta, fasta -> [ meta + [ 'haplotype': 2 ], fasta ] }
        .set { ch_gfastats_maternal_in }

    GFASTATS(
        ch_gfastats_paternal_in.mix(ch_gfastats_maternal_in),
        'fasta',
        '',
        '',
        [[],[]],
        [[],[]],
        [[],[]],
        [[],[]]
    )
    ch_versions = ch_versions.mix(GFASTATS.out.versions)

    emit:
    assembled_haplotypes = GFASTATS.out.assembly // channel: [Â val(meta), path(fasta) ]
    versions = ch_versions                       // channel: [ versions.yml ]
}
