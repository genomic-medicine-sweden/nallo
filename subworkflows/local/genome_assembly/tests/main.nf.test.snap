{
    "two children with two parents, trio-binning mode": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true,
                            "haplotype": 1
                        },
                        "HG002_haplotype_1.fasta.gz:md5,27b5d0dc68bd7cd9d0b7a2fea33ca1e4"
                    ],
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true,
                            "haplotype": 2
                        },
                        "HG002_haplotype_2.fasta.gz:md5,62b8128e71e55bef9c06711f29fe2302"
                    ],
                    [
                        {
                            "id": "HG003",
                            "n_files": 1,
                            "relationship": "father",
                            "children": [
                                "HG002",
                                "second_child"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 1
                        },
                        "HG003_haplotype_1.fasta.gz:md5,7cf7379bc90d4cabea1ae4bad56d9e6d"
                    ],
                    [
                        {
                            "id": "HG003",
                            "n_files": 1,
                            "relationship": "father",
                            "children": [
                                "HG002",
                                "second_child"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 2
                        },
                        "HG003_haplotype_2.fasta.gz:md5,372bd5f78974ad2e394130d790ee9ca0"
                    ],
                    [
                        {
                            "id": "HG004",
                            "n_files": 1,
                            "relationship": "mother",
                            "children": [
                                "HG002",
                                "second_child"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 1
                        },
                        "HG004_haplotype_1.fasta.gz:md5,e2eaf0517d02d226ab3799cb709216bc"
                    ],
                    [
                        {
                            "id": "HG004",
                            "n_files": 1,
                            "relationship": "mother",
                            "children": [
                                "HG002",
                                "second_child"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 2
                        },
                        "HG004_haplotype_2.fasta.gz:md5,3cbbafd35aaf785fa3cb2604c38f6f2f"
                    ],
                    [
                        {
                            "id": "second_child",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true,
                            "haplotype": 1
                        },
                        "second_child_haplotype_1.fasta.gz:md5,27b5d0dc68bd7cd9d0b7a2fea33ca1e4"
                    ],
                    [
                        {
                            "id": "second_child",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true,
                            "haplotype": 2
                        },
                        "second_child_haplotype_2.fasta.gz:md5,62b8128e71e55bef9c06711f29fe2302"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true
                        },
                        "HG002_other.fastq.gz:md5,f9d9d405021d7a64eb59929e90aed241"
                    ],
                    [
                        {
                            "id": "second_child",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true
                        },
                        "second_child_other.fastq.gz:md5,f9d9d405021d7a64eb59929e90aed241"
                    ]
                ],
                "2": [
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,f7a9ef5f86ecf68fab47c66d96fee88e",
                    "versions.yml:md5,f7a9ef5f86ecf68fab47c66d96fee88e"
                ],
                "assembled_haplotypes": [
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true,
                            "haplotype": 1
                        },
                        "HG002_haplotype_1.fasta.gz:md5,27b5d0dc68bd7cd9d0b7a2fea33ca1e4"
                    ],
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true,
                            "haplotype": 2
                        },
                        "HG002_haplotype_2.fasta.gz:md5,62b8128e71e55bef9c06711f29fe2302"
                    ],
                    [
                        {
                            "id": "HG003",
                            "n_files": 1,
                            "relationship": "father",
                            "children": [
                                "HG002",
                                "second_child"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 1
                        },
                        "HG003_haplotype_1.fasta.gz:md5,7cf7379bc90d4cabea1ae4bad56d9e6d"
                    ],
                    [
                        {
                            "id": "HG003",
                            "n_files": 1,
                            "relationship": "father",
                            "children": [
                                "HG002",
                                "second_child"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 2
                        },
                        "HG003_haplotype_2.fasta.gz:md5,372bd5f78974ad2e394130d790ee9ca0"
                    ],
                    [
                        {
                            "id": "HG004",
                            "n_files": 1,
                            "relationship": "mother",
                            "children": [
                                "HG002",
                                "second_child"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 1
                        },
                        "HG004_haplotype_1.fasta.gz:md5,e2eaf0517d02d226ab3799cb709216bc"
                    ],
                    [
                        {
                            "id": "HG004",
                            "n_files": 1,
                            "relationship": "mother",
                            "children": [
                                "HG002",
                                "second_child"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 2
                        },
                        "HG004_haplotype_2.fasta.gz:md5,3cbbafd35aaf785fa3cb2604c38f6f2f"
                    ],
                    [
                        {
                            "id": "second_child",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true,
                            "haplotype": 1
                        },
                        "second_child_haplotype_1.fasta.gz:md5,27b5d0dc68bd7cd9d0b7a2fea33ca1e4"
                    ],
                    [
                        {
                            "id": "second_child",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true,
                            "haplotype": 2
                        },
                        "second_child_haplotype_2.fasta.gz:md5,62b8128e71e55bef9c06711f29fe2302"
                    ]
                ],
                "two_parent_children": [
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true
                        },
                        "HG002_other.fastq.gz:md5,f9d9d405021d7a64eb59929e90aed241"
                    ],
                    [
                        {
                            "id": "second_child",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true
                        },
                        "second_child_other.fastq.gz:md5,f9d9d405021d7a64eb59929e90aed241"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,f7a9ef5f86ecf68fab47c66d96fee88e",
                    "versions.yml:md5,f7a9ef5f86ecf68fab47c66d96fee88e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-07T12:08:03.62209152"
    },
    "child with two parents, trio-binning mode": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true,
                            "haplotype": 1
                        },
                        "HG002_haplotype_1.fasta.gz:md5,27b5d0dc68bd7cd9d0b7a2fea33ca1e4"
                    ],
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true,
                            "haplotype": 2
                        },
                        "HG002_haplotype_2.fasta.gz:md5,62b8128e71e55bef9c06711f29fe2302"
                    ],
                    [
                        {
                            "id": "HG003",
                            "n_files": 1,
                            "relationship": "father",
                            "children": [
                                "HG002"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 1
                        },
                        "HG003_haplotype_1.fasta.gz:md5,7cf7379bc90d4cabea1ae4bad56d9e6d"
                    ],
                    [
                        {
                            "id": "HG003",
                            "n_files": 1,
                            "relationship": "father",
                            "children": [
                                "HG002"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 2
                        },
                        "HG003_haplotype_2.fasta.gz:md5,372bd5f78974ad2e394130d790ee9ca0"
                    ],
                    [
                        {
                            "id": "HG004",
                            "n_files": 1,
                            "relationship": "mother",
                            "children": [
                                "HG002"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 1
                        },
                        "HG004_haplotype_1.fasta.gz:md5,e2eaf0517d02d226ab3799cb709216bc"
                    ],
                    [
                        {
                            "id": "HG004",
                            "n_files": 1,
                            "relationship": "mother",
                            "children": [
                                "HG002"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 2
                        },
                        "HG004_haplotype_2.fasta.gz:md5,3cbbafd35aaf785fa3cb2604c38f6f2f"
                    ]
                ],
                "1": [
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,f7a9ef5f86ecf68fab47c66d96fee88e",
                    "versions.yml:md5,f7a9ef5f86ecf68fab47c66d96fee88e"
                ],
                "assembled_haplotypes": [
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true,
                            "haplotype": 1
                        },
                        "HG002_haplotype_1.fasta.gz:md5,27b5d0dc68bd7cd9d0b7a2fea33ca1e4"
                    ],
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": true,
                            "haplotype": 2
                        },
                        "HG002_haplotype_2.fasta.gz:md5,62b8128e71e55bef9c06711f29fe2302"
                    ],
                    [
                        {
                            "id": "HG003",
                            "n_files": 1,
                            "relationship": "father",
                            "children": [
                                "HG002"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 1
                        },
                        "HG003_haplotype_1.fasta.gz:md5,7cf7379bc90d4cabea1ae4bad56d9e6d"
                    ],
                    [
                        {
                            "id": "HG003",
                            "n_files": 1,
                            "relationship": "father",
                            "children": [
                                "HG002"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 2
                        },
                        "HG003_haplotype_2.fasta.gz:md5,372bd5f78974ad2e394130d790ee9ca0"
                    ],
                    [
                        {
                            "id": "HG004",
                            "n_files": 1,
                            "relationship": "mother",
                            "children": [
                                "HG002"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 1
                        },
                        "HG004_haplotype_1.fasta.gz:md5,e2eaf0517d02d226ab3799cb709216bc"
                    ],
                    [
                        {
                            "id": "HG004",
                            "n_files": 1,
                            "relationship": "mother",
                            "children": [
                                "HG002"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": true,
                            "haplotype": 2
                        },
                        "HG004_haplotype_2.fasta.gz:md5,3cbbafd35aaf785fa3cb2604c38f6f2f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,f7a9ef5f86ecf68fab47c66d96fee88e",
                    "versions.yml:md5,f7a9ef5f86ecf68fab47c66d96fee88e"
                ]
            },
            [
                [
                    {
                        "id": "HG002",
                        "n_files": 1,
                        "relationship": "child",
                        "children": [
                            
                        ],
                        "family_id": "NIST",
                        "two_parents": true,
                        "haplotype": 1
                    },
                    "HG002_haplotype_1.fasta.gz:md5,27b5d0dc68bd7cd9d0b7a2fea33ca1e4"
                ],
                [
                    {
                        "id": "HG002",
                        "n_files": 1,
                        "relationship": "child",
                        "children": [
                            
                        ],
                        "family_id": "NIST",
                        "two_parents": true,
                        "haplotype": 2
                    },
                    "HG002_haplotype_2.fasta.gz:md5,62b8128e71e55bef9c06711f29fe2302"
                ],
                [
                    {
                        "id": "HG003",
                        "n_files": 1,
                        "relationship": "father",
                        "children": [
                            "HG002"
                        ],
                        "family_id": "NIST",
                        "has_other_parent": true,
                        "haplotype": 1
                    },
                    "HG003_haplotype_1.fasta.gz:md5,7cf7379bc90d4cabea1ae4bad56d9e6d"
                ],
                [
                    {
                        "id": "HG003",
                        "n_files": 1,
                        "relationship": "father",
                        "children": [
                            "HG002"
                        ],
                        "family_id": "NIST",
                        "has_other_parent": true,
                        "haplotype": 2
                    },
                    "HG003_haplotype_2.fasta.gz:md5,372bd5f78974ad2e394130d790ee9ca0"
                ],
                [
                    {
                        "id": "HG004",
                        "n_files": 1,
                        "relationship": "mother",
                        "children": [
                            "HG002"
                        ],
                        "family_id": "NIST",
                        "has_other_parent": true,
                        "haplotype": 1
                    },
                    "HG004_haplotype_1.fasta.gz:md5,e2eaf0517d02d226ab3799cb709216bc"
                ],
                [
                    {
                        "id": "HG004",
                        "n_files": 1,
                        "relationship": "mother",
                        "children": [
                            "HG002"
                        ],
                        "family_id": "NIST",
                        "has_other_parent": true,
                        "haplotype": 2
                    },
                    "HG004_haplotype_2.fasta.gz:md5,3cbbafd35aaf785fa3cb2604c38f6f2f"
                ]
            ],
            14
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-07T11:56:04.505705832"
    },
    "one sample, hifi-only mode": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_fastq",
                            "n_files": 1,
                            "relationship": "unknown",
                            "haplotype": 1
                        },
                        "test_fastq_haplotype_1.fasta.gz:md5,a6c5cbbeff6da2f9bb2e16639f216f25"
                    ],
                    [
                        {
                            "id": "test_fastq",
                            "n_files": 1,
                            "relationship": "unknown",
                            "haplotype": 2
                        },
                        "test_fastq_haplotype_2.fasta.gz:md5,7c89a495b047249f49c4cd47841b71eb"
                    ]
                ],
                "1": [
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659"
                ],
                "assembled_haplotypes": [
                    [
                        {
                            "id": "test_fastq",
                            "n_files": 1,
                            "relationship": "unknown",
                            "haplotype": 1
                        },
                        "test_fastq_haplotype_1.fasta.gz:md5,a6c5cbbeff6da2f9bb2e16639f216f25"
                    ],
                    [
                        {
                            "id": "test_fastq",
                            "n_files": 1,
                            "relationship": "unknown",
                            "haplotype": 2
                        },
                        "test_fastq_haplotype_2.fasta.gz:md5,7c89a495b047249f49c4cd47841b71eb"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659"
                ]
            },
            [
                "\"GENOME_ASSEMBLY:HIFIASM\":",
                "    hifiasm: 0.25.0-r726",
                "\"GENOME_ASSEMBLY:GFASTATS\":",
                "    gfastats: 1.3.10",
                "\"GENOME_ASSEMBLY:GFASTATS\":",
                "    gfastats: 1.3.10"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-07T10:53:15.711017842"
    },
    "child with one parent, trio-binning mode": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": false,
                            "haplotype": 1
                        },
                        "HG002_haplotype_1.fasta.gz:md5,a6c5cbbeff6da2f9bb2e16639f216f25"
                    ],
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": false,
                            "haplotype": 2
                        },
                        "HG002_haplotype_2.fasta.gz:md5,7c89a495b047249f49c4cd47841b71eb"
                    ],
                    [
                        {
                            "id": "HG003",
                            "n_files": 1,
                            "relationship": "father",
                            "children": [
                                "HG002"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": false,
                            "haplotype": 1
                        },
                        "HG003_haplotype_1.fasta.gz:md5,7cf7379bc90d4cabea1ae4bad56d9e6d"
                    ],
                    [
                        {
                            "id": "HG003",
                            "n_files": 1,
                            "relationship": "father",
                            "children": [
                                "HG002"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": false,
                            "haplotype": 2
                        },
                        "HG003_haplotype_2.fasta.gz:md5,372bd5f78974ad2e394130d790ee9ca0"
                    ]
                ],
                "1": [
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659"
                ],
                "assembled_haplotypes": [
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": false,
                            "haplotype": 1
                        },
                        "HG002_haplotype_1.fasta.gz:md5,a6c5cbbeff6da2f9bb2e16639f216f25"
                    ],
                    [
                        {
                            "id": "HG002",
                            "n_files": 1,
                            "relationship": "child",
                            "children": [
                                
                            ],
                            "family_id": "NIST",
                            "two_parents": false,
                            "haplotype": 2
                        },
                        "HG002_haplotype_2.fasta.gz:md5,7c89a495b047249f49c4cd47841b71eb"
                    ],
                    [
                        {
                            "id": "HG003",
                            "n_files": 1,
                            "relationship": "father",
                            "children": [
                                "HG002"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": false,
                            "haplotype": 1
                        },
                        "HG003_haplotype_1.fasta.gz:md5,7cf7379bc90d4cabea1ae4bad56d9e6d"
                    ],
                    [
                        {
                            "id": "HG003",
                            "n_files": 1,
                            "relationship": "father",
                            "children": [
                                "HG002"
                            ],
                            "family_id": "NIST",
                            "has_other_parent": false,
                            "haplotype": 2
                        },
                        "HG003_haplotype_2.fasta.gz:md5,372bd5f78974ad2e394130d790ee9ca0"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,2c13b2225c94411809dff121aa467164",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659",
                    "versions.yml:md5,34637a8dde7b25a7df56d1ab50a8a659"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-07T12:15:00.511059735"
    }
}